helix-term/src/commands.rs

```diff
@@ -5060,15 +5060,31 @@ fn format_selections(cx: &mut Context) {
         )
         .unwrap();
 
-    let edits = tokio::task::block_in_place(|| helix_lsp::block_on(future))
-        .ok()
-        .flatten()
-        .unwrap_or_default();
-
-    let transaction =
-        helix_lsp::util::generate_transaction_from_edits(doc.text(), edits, offset_encoding);
-
-    doc.apply(&transaction, view_id);
+    let text = doc.text().clone();
+    let doc_id = doc.id();
+    let doc_version = doc.version();
+
+    tokio::spawn(async move {
+        match future.await {
+            Ok(Some(res)) => {
+                let transaction =
+                    helix_lsp::util::generate_transaction_from_edits(&text, res, offset_encoding);
+                job::dispatch(move |editor, _compositor| {
+                    let Some(doc) = editor.document_mut(doc_id) else {
+                        return;
+                    };
+                    // Updating a desynced document causes problems with applying the transaction
+                    if doc.version() != doc_version {
+                        return;
+                    }
```

Comment:
👍 nice catch, this would've been the main thing I would've been concerned about with #13238